File: January2021/C++/ValidParentheses.cpp
/**
 * 
 * @author ssrbazpur
 * 
 * */

class Solution
{
public:
    bool isValid(string s1)
    {
        stack<char> s;
        bool ans = true;
        for (char a : s1)
        {
            if (s.empty())
            {
                s.push(a);
            }
            else
            {
                char t = s.top();
                if (a == ']' && t == '[')
                {
                    s.pop();
                }
                else if (a == ')' && t == '(')
                {
                    s.pop();
                }
                else if (a == '}' && t == '{')
                {
                    s.pop();
                }
                else
                {
                    s.push(a);
                }
            }
        }

        return s.empty();
    }
};

File: March2022/Java/ValidParentheses.java

class Solution {
    // TC O(len(s))
    // SC O(len(s))
    public boolean isValid(String s) {
        Stack<Character> st =new Stack<Character>();
        for(char c: s.toCharArray()){
            if(st.empty()){
                st.push(c);
            } else{
                char topChar = st.peek();
                if(topChar == '(' && c ==')'){
                    st.pop();
                } else  if(topChar == '{' && c =='}'){
                    st.pop();
                } else  if(topChar == '[' && c ==']'){
                    st.pop();
                } else{
                    st.push(c);
                }
            }
        }
        return st.empty();
    }
}